### ISSUE TITLE
Non-Seekable SubReadStream should not support the Length property.

### ISSUE THREAD

**Issue by JimBobSquarePants** [dotnet/runtime#63040]
### Description

According to the documentation for [Stream.Length#Exceptions](https://docs.microsoft.com/en-us/dotnet/api/system.io.stream.length?redirectedfrom=MSDN&view=net-6.0#exceptions) a `NotSupportedException` should be thrown when .

>NotSupportedException
A class derived from Stream does not support seeking.

### Reproduction Steps

The following test code should be enough to reproduce. ZipArchive sample taken from https://github.com/SixLabors/ImageSharp/issues/1903

```c#
using var zipFile = new ZipArchive(new MemoryStream(
    new byte[]
    {
        0x50, 0x4B, 0x03, 0x04, 0x0A, 0x00, 0x00, 0x00, 0x00, 0x00, 0x77, 0xAF,
        0x94, 0x53, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
        0x00, 0x00, 0x07, 0x00, 0x00, 0x00, 0x6D, 0x79, 0x73, 0x74, 0x65, 0x72,
        0x79, 0x50, 0x4B, 0x01, 0x02, 0x3F, 0x00, 0x0A, 0x00, 0x00, 0x00, 0x00,
        0x00, 0x77, 0xAF, 0x94, 0x53, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
        0x00, 0x00, 0x00, 0x00, 0x00, 0x07, 0x00, 0x24, 0x00, 0x00, 0x00, 0x00,
        0x00, 0x00, 0x00, 0x20, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x6D,
        0x79, 0x73, 0x74, 0x65, 0x72, 0x79, 0x0A, 0x00, 0x20, 0x00, 0x00, 0x00,
        0x00, 0x00, 0x01, 0x00, 0x18, 0x00, 0x46, 0x82, 0xFF, 0x91, 0x27, 0xF6,
        0xD7, 0x01, 0x55, 0xA1, 0xF9, 0x91, 0x27, 0xF6, 0xD7, 0x01, 0x55, 0xA1,
        0xF9, 0x91, 0x27, 0xF6, 0xD7, 0x01, 0x50, 0x4B, 0x05, 0x06, 0x00, 0x00,
        0x00, 0x00, 0x01, 0x00, 0x01, 0x00, 0x59, 0x00, 0x00, 0x00, 0x25, 0x00,
        0x00, 0x00, 0x00, 0x00
    }));
using Stream stream = zipFile.Entries[0].Open();
Assert.Throws<NotSupportedException>(() => stream.Length);
```

### Expected behavior

The stream should throw a `NotSupportedException`

### Actual behavior

The stream returns the length.

### Configuration

 .NET 6 Any OS and architecture.

**Comment by mgravell** [comment]
Counter opinion (entirely subjective): if the length is knowable, it is IMO fine and desirable to report it. For example, a bounded once-only stream may be known-length yet non-seekable, and in that case: it remains desirable to report the length, for use in things like headers / prefixes, to avoid an extra buffer/count step. If the consuming code cares about seekability, it should test CanSeek.

In other words, I'm saying that *personally* I'd interpret this as a "may", not a "must" (throw, in the circumstances).

**Comment by JimBobSquarePants** [comment]
The [CanSeek](https://docs.microsoft.com/en-us/dotnet/api/system.io.stream.canseek?view=net-6.0#remarks) docs agree with you @mgravell (Thanks @scalablecory). 

I'm fine with closing this then and raising something against the docs repo instead. `Length` reads (to me anyway) as explicit.

**Comment by adamsitnik** [comment]
@JimBobSquarePants would it be enough if we have updated the docs in the following way?

```diff
- A class derived from Stream does not support seeking.
+ A class derived from Stream does not support seeking and the length is unknown.
```

**Comment by JimBobSquarePants** [comment]
@adamsitnik Yeah that's much better! 👍

**Comment by Sergio0694** [comment]
Possibly related to https://github.com/dotnet/runtime/issues/43448 😊

**Comment by antonfirsov** [comment]
> [Seeking to any location beyond the length of the stream is supported.](https://docs.microsoft.com/en-us/dotnet/api/system.io.stream.position?view=net-6.0#remarks)

I wonder what are detailed semantics for this. For example is it a requirement for `Stream.Position` to roundtrip?

```C#
stream.Position = stream.Length * 2;
Assert.Equal(stream.Length * 2, stream.Position);
```

Or going more extreme:
```C#
stream.Position = long.MaxValue;
Assert.Equal(long.MaxValue, stream.Position);
```

**Comment by zmj** [comment]
> For example, a bounded once-only stream may be known-length yet non-seekable, and in that case: it remains desirable to report the length, for use in things like headers / prefixes, to avoid an extra buffer/count step. If the consuming code cares about seekability, it should test CanSeek.

On the consuming side of Streams like this, it would be nice to have a non-throwing, non-CanSeek way to try to get a length.
